File: MVP/app/src/main/java/com/example/mvp/main/favproduct/view/FavActivity.java
package com.example.mvp.main.favproduct.view;

import android.os.Bundle;

import androidx.activity.EdgeToEdge;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.Observer;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.example.mvp.R;
import com.example.mvp.main.db.ProductLocalDataSourceImp;
import com.example.mvp.main.favproduct.FavContract;
import com.example.mvp.main.favproduct.presenter.FavPresenter;
import com.example.mvp.main.model.Product;
import com.example.mvp.main.model.ProductRepositoryImp;
import com.example.mvp.main.network.ProductRemoteDataSourceImp;

import java.util.List;

public class FavActivity extends AppCompatActivity implements FavContract.view,onRemoveClick {

    RecyclerView favRecview ;

    FavProductAdapter adapter ;
    FavPresenter presenter ;

    LinearLayoutManager manager ;

    LiveData<List<Product>> products ;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_fav);

        presenter = new FavPresenter(new ProductRepositoryImp(ProductRemoteDataSourceImp.getInstance(),ProductLocalDataSourceImp.getInstance(this)));

        products = presenter.getStored();
        initUI();
        manager = new LinearLayoutManager(this);
        favRecview.setLayoutManager(manager);


        products.observe(this, new Observer<List<Product>>() {
            @Override
            public void onChanged(List<Product> products) {
            adapter = new FavProductAdapter(FavActivity.this,products,favRecview,FavActivity.this);
            favRecview.setAdapter(adapter);
            }
        });

    }

    @Override
    public void onRemClick(Product product) {
    presenter.removeFav(product);
    }


    @Override
    public void initUI() {
        favRecview = findViewById(R.id.recyclerView);
    }
}

File: MVVM/app/src/main/java/com/example/mvvm/favproducts/FavActivity.kt
package com.example.mvvm.favproducts

import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.mvvm.R
import com.example.mvvm.allproducts.AllProductsViewModel
import com.example.mvvm.databinding.ActivityFavBinding
import com.example.mvvm.model.db.ProductLocalDataSourceImp
import com.example.mvvm.model.network.ProductRemoteDataSourceImp
import com.example.mvvm.model.repo.ProductRepoImp

class FavActivity : AppCompatActivity() {

    lateinit var binding: ActivityFavBinding
    lateinit var favProductViewModel: FavProductViewModel
    lateinit var favProductAdapter: FavProductAdapter
    lateinit var myManager : LinearLayoutManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        binding = ActivityFavBinding.inflate(layoutInflater)
        setContentView(binding.root)

        myManager = LinearLayoutManager(this)


        favProductViewModel = ViewModelProvider(this, FavProductsViewModelFactory
            (ProductRepoImp.getInstance(ProductLocalDataSourceImp.getInstance(this), ProductRemoteDataSourceImp)))
            .get(FavProductViewModel::class.java)

        favProductAdapter = FavProductAdapter()
        {
         favProductViewModel.deleteFav(it)
        }.apply { submitList(emptyList()) }

        binding.recyclerView.apply {
            adapter = favProductAdapter
            layoutManager =myManager
        }

        favProductViewModel.productList.observe(this) {
            favProductAdapter.submitList(it)
        }
        favProductViewModel.getAllFavProducts()


    }

}